resources/js/tableofcontent.js

```diff
@@ -1,3 +1,63 @@
+function isHeaderElement(elem) {
+    return elem.nodeName.match(/^H\d+$/) && elem.textContent;
+}
+
+function createHeaderData(elem, i) {
+    const headerText = elem.textContent.trim().replace(/"/g, '\\"');
+    const level = parseInt(elem.nodeName.substr(1));
+    const anchor = `kiwix-toc-${i}`;
+
+    const anchorElem = document.createElement("a");
+    anchorElem.id = anchor;
+
+    /* Mark header content with something we can reference. */
+    elem.insertAdjacentElement("afterbegin", anchorElem);
+
+    return { headerEntry: { text: headerText, anchor, table: [] }, level };
+}
+
+function getDOMHeadersPreorder(elem, array)
+{
+    if (isHeaderElement(elem))
+        array.push(elem);
+
+    Array.from(elem.children).forEach(child => getDOMHeadersPreorder(child, array));
+}
+
+function createTOCFromHeaders(headers, toc)
+{
+    /* Stack to keep track of parent header levels. */
+    const levelStack = [];
+    headers.forEach(function(elem, i) {
+        const { headerEntry, level } = createHeaderData(elem, i);
+
+        /* Pop until the correct level for insertion. */
+        while (levelStack.length !== 0 && level <= levelStack[levelStack.length - 1].level)
+            levelStack.pop();
+
+        if (levelStack.length === 0)
+            toc.table.push(headerEntry);
+        else
+            levelStack[levelStack.length - 1].table.push(headerEntry);
+
+        levelStack.push({ table: headerEntry.table, level });
+    }); 
+}
+
+function tocJSON()
+{
+    const toc = { url: window.location.href.replace(location.hash,""), table: [] };
+
+    if (document.body !== undefined)
+    {
+        const headers = [];
+        getDOMHeadersPreorder(document.body, headers);
+        createTOCFromHeaders(headers, toc);
```

Comment:
Why not `toc.table = createTOC(headers);`?

Reply:
Done